lib/Sema/TypeCheckDecl.cpp

```diff
@@ -8133,14 +8194,48 @@ void TypeChecker::addImplicitConstructors(NominalTypeDecl *decl) {
   bool FoundMemberwiseInitializedProperty = false;
   bool SuppressDefaultInitializer = false;
   bool SuppressMemberwiseInitializer = false;
+  bool FoundSynthesizedInit = false;
   bool FoundDesignatedInit = false;
   decl->setAddedImplicitInitializers();
+
+  // Before we look for constructors, we need to make sure that all synthesized
+  // initializers are properly synthesized.
+  //
+  // NOTE: Lookups of synthesized initializers MUST come after
+  //       decl->setAddedImplicitInitializers() in case synthesis requires
+  //       protocol conformance checking, which might be recursive here.
+  DeclName synthesizedInitializers[1] = {
+    // init(from:) is synthesized by derived conformance to Decodable.
+    DeclName(Context, DeclBaseName(Context.Id_init),
+             (Identifier[1]) { Context.Id_from })
+  };
```

Comment:
Unfortunately, we do have cases where we appear to rely on this (I believe somewhere deep in the `String` implementation in the stdlib). The first thing I tried was simply to prevent any implicit initializer from preventing synthesis, but the `stdlib` couldn't compile because it was crashing in compiling `String` (or one of the related types).

I can investigate this further to give you more info, but this was to avoid that from happening. ☹️